crates/bevy_ecs/macros/src/component.rs

```diff
@@ -138,9 +142,14 @@ pub fn derive_component(input: TokenStream) -> TokenStream {
                 storages: &mut #bevy_ecs_path::storage::Storages,
                 required_components: &mut #bevy_ecs_path::component::RequiredComponents,
                 inheritance_depth: u16,
+                recursion_check_stack: &mut Vec<#bevy_ecs_path::component::ComponentId>
             ) {
+                #bevy_ecs_path::component::enforce_no_required_components_recursion(components, recursion_check_stack);
```

Comment:
I think doing this check before the `push` means you have one more level of recursion than is necessary.  

Like, if you register a component that requires itself, then the first time this is called the stack will be empty, and the second time it will have one element that it binds to `requiree` but an empty `check` list.  So it will pass on the second call, even though it had enough information to detect the error!  

The simplest thing to do might be to pass `self_id` to `enforce_no_required_components_recursion` so that it doesn't need to do `split_last()`.